packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart

```diff
@@ -0,0 +1,316 @@
+            DefaultCloudWatchLogStreamProvider(
+              logGroupName: pluginConfig.logGroupName,
+              client: CloudWatchLogsClient(
+                region: pluginConfig.region,
+                credentialsProvider: credentialsProvider,
+              ),
+            );
+
+  final CloudWatchLoggerPluginConfiguration _pluginConfig;
+  final CloudWatchLogsClient _clinet;
+  final CloudWatchLogStreamProvider _logStreamProvider;
+  final QueuedItemStore _logStore;
+  bool _syncing;
+  bool _enabled;
+  StoppableTimer? _timer;
+  RemoteLoggingConstraintProvider? _remoteLoggingConstraintProvider;
+
+  set remoteLoggingConstraintProvider(
+    RemoteLoggingConstraintProvider remoteProvider,
+  ) {
+    if (_remoteLoggingConstraintProvider != null) {
+      logger.error(
+        'remoteLoggingConstraintProvider is already configured.',
+      );
+      return;
+    }
+    _remoteLoggingConstraintProvider = remoteProvider;
+  }
+
+  Future<void> _startSyncingIfNotInProgress() async {
+    Future<void> startSyncing() async {
+      String logStream;
+      try {
+        logStream = await _logStreamProvider.logStream;
+      } on Exception catch (e) {
+        if (await _logStore.isFull(_pluginConfig.localStoreMaxSizeInMB)) {
+          _logStore.clear();
+          logger.warn(
+            'Reached local store max size of: '
+            '${_pluginConfig.localStoreMaxSizeInMB}.Hence logs are deleted from '
+            'local store.',
+          );
+        }
+        logger.error('Failed to create CloudWatch log stream', e);
+        return;
+      }
+
+      final batches = await _getLogBatchesToSync();
+      for (var i = 0; i < batches.length; i++) {
+        try {
+          final response = await _sendToCloudWatch(batches[i].$2, logStream);
+          if (response.rejectedLogEventsInfo?.tooNewLogEventStartIndex !=
+              null) {
+            break;
+          } else {
+            _logStore.deleteItems(batches[i].$1);
+          }
+        } on Exception catch (e) {
+          if (await _logStore.isFull(_pluginConfig.localStoreMaxSizeInMB)) {
+            _logStore.deleteItems(batches[i].$1);
+            logger.warn(
+              'Reached local store max size of: '
+              '${_pluginConfig.localStoreMaxSizeInMB}.Hence logs are deleted '
+              'from local store.',
+            );
+          }
+          logger.error('Failed to sync batched logs to CloudWatch', e);
+        }
+      }
+    }
+
+    if (!_syncing) {
+      _syncing = true;
+      try {
+        await startSyncing();
+      } on Exception catch (e) {
+        logger.error('Failed to sync logs to CloudWatch.', e);
+      } finally {
+        _syncing = false;
+      }
+    }
+  }
+
+  void _onTimerError(Object e) {
+    logger.error('Failed to sync logs to CloudWatch.', e);
+  }
+
+  Future<LoggingConstraint> _getLoggingConstraint() async {
+    final result = await _remoteLoggingConstraintProvider?.loggingConstraint;
```

Comment:
the remote provider will do the caching.

Reply:
Can you add that to the docs for the class then if it's the expectation of developers to implement their own caching logic?

Reply:
If they don't, then this will run an HTTP call or some other async work for every log message which is insane.

Even an empty `Future` incurs a penalty to be running for every log message.